automap.c

```diff
@@ -832,78 +1814,231 @@ fam_get(FAMObject *self, PyObject *args)
+        i++;                                                   \
+    }                                                          \
+}                                                              \
+else {                                                         \
+    for (; i < keys_size; i++) {                               \
+        char_type* v = (char_type*)PyArray_GETPTR1(a, i);      \
+        p = get_end_func(v, dt_size);                          \
+        if (insert_func(fam, v, p-v, i, -1)) {                 \
+            goto error;                                        \
+        }                                                      \
+    }                                                          \
+}                                                              \
 
 static PyObject *
 fam_new(PyTypeObject *cls, PyObject *args, PyObject *kwargs)
 {
+    // NOTE: The original fam_new used to be able to provide a same reference back if a fam was in the args; this is tricky now that we have fam_init
+    FAMObject *self = (FAMObject *)cls->tp_alloc(cls, 0);
+    if (!self) {
+        return NULL;
+    }
+    self->table = NULL;
+    self->keys = NULL;
+    self->key_buffer = NULL;
+    self->keys_size = 0;
+
+    return (PyObject*)self;
+}
+
+
+// Initialize an allocated FAMObject. Returns 0 on success, -1 on error.
+int
+fam_init(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    PyTypeObject* cls = Py_TYPE(self); // borrowed ref
     const char *name = cls->tp_name;
+    FAMObject* fam = (FAMObject*)self;
+
     if (kwargs) {
         PyErr_Format(PyExc_TypeError, "%s takes no keyword arguments", name);
-        return NULL;
+        return -1;
     }
+
+    int keys_array_type = KAT_LIST; // default, will override if necessary
+
     PyObject *keys = NULL;
+    Py_ssize_t keys_size = 0;
+
     if (!PyArg_UnpackTuple(args, name, 0, 1, &keys)) {
-        return NULL;
+        return -1;
     }
+
     if (!keys) {
         keys = PyList_New(0);
     }
     else if (PyObject_TypeCheck(keys, &FAMType)) {
-        return (PyObject *)copy(cls, (FAMObject *)keys);
+        // Use `keys` as old, `self` as new, and fill from old to new. This returns the same error codes as this function.
+        return copy_to_new(cls, (FAMObject*)keys, fam);
     }
-    else {
+    else if (PyArray_Check(keys)) {
+        PyArrayObject *a = (PyArrayObject *)keys;
+        if (PyArray_NDIM(a) != 1) {
+            PyErr_SetString(PyExc_TypeError, "Arrays must be 1-dimensional");
+            return -1;
+        }
+        int array_t = PyArray_TYPE(a);
+        if (cls != &AMType &&
+                (PyTypeNum_ISINTEGER(array_t) // signed and unsigned
+                || PyTypeNum_ISFLOAT(array_t)
+                || PyTypeNum_ISFLEXIBLE(array_t))
+            ){
+            if ((PyArray_FLAGS(a) & NPY_ARRAY_WRITEABLE)) {
+                PyErr_Format(PyExc_TypeError, "Arrays must be immutable when given to a %s", name);
```

Comment:
This is not what I would expect the behavior to be. I would imagine it simply checks if the array is immutable. If it is not, then just proceed using the old code.

I think it's unexpected (and undesirable) behavior to fail instantiating a `FrozenAutoMap` when given a numpy array that is not immutable. This seems like the responsibility of the client, not automap.

Reply:
Are you sure? As the performance characteristics are far different between using immutable arrays and converting those arrays to lists, I think it is important to only support one way of creating a `FAM` with supported array types. Otherwise, one could silently fall into using the old code simply because the client forgot to set the array as immutable.

Reply:
I do feel rather strongly about it, as my expectations for `automap` is that it's designed to be a drop-in replacement for sets.

In other words:
`AutoMap` <=> `set`
`FrozenAutoMap` <=> `frozenset`

So, just like you can do `frozenset(mutable_numpy_arr)`, I would expect you could also do `FrozenAutoMap(mutable_numpy_arr)`

Reply:
But why would it benefit the caller to support sub-optimal usage from a mutable array? The caller probably does not want that, and would rather get an error telling them to make the array immutable. If we really wanted to provide this convenience, we could make a copy of the array ourselves and set it to be immutable, but that would also introduce a silent performance penalty (but would probably still deliver better performance than converting the array to PyObjects).

Reply:
I guess it's my opinion that it's not `automap`'s call to make. This is a great optimization, but it seems like an anti-pattern to force callers to modify their input in order to take advantage of this optimization

I think it would be better to have this be a `PerformanceWarning`, that lets users know about the overhead of using mutable arrays.

I feel strongly that it's the client's responsibility to opt-in, rather than automap's responsibility to disallow mutable arrays entirely

Reply:
Issuing a warning at least makes the undesirable usage not silent! I see your point that it is unconventional for the container to reject such input. But I also do not see `FAM` as having much general-purpose usage outside of StaticFrame... in the context of StaticFrame, I want a failure if I accidentally give a `FAM` a mutable array!